packages/options/src/option.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * An implementation of the Rust Option type in JavaScript.
+ * It can be one of the following:
+ * - <code>{@link Some}<T></code>: Meaning there is a value of type T.
+ * - <code>{@link None}</code>: Meaning there is no value.
+ */
+export type Option<T> = Some<T> | None;
+
+/**
+ * Defines a looser type that can be used when serializing an {@link Option}.
+ * This allows us to pass null or the Option value directly whilst still
+ * supporting the Option type for use-cases that need more type safety.
+ */
+export type OptionOrNullable<T> = Option<T> | T | null;
+
+/**
+ * Represents an option of type `T` that has a value.
+ *
+ * @see {@link Option}
+ */
+export type Some<T> = { __option: 'Some'; value: T };
+
+/**
+ * Represents an option of type `T` that has no value.
+ *
+ * @see {@link Option}
+ */
+export type None = { __option: 'None' };
+
+/**
+ * Creates a new {@link Option} of type `T` that has a value.
+ *
+ * @see {@link Option}
+ */
+export const some = <T>(value: T): Option<T> => ({ __option: 'Some', value });
+
+/**
+ * Creates a new {@link Option} of type `T` that has no value.
+ *
+ * @see {@link Option}
+ */
+export const none = <T>(): Option<T> => ({ __option: 'None' });
```

Comment:
How would you feel about literally returning the same (frozen) object?
```suggestion
const NONE = Object.freeze({ __option: 'None' });
export const none = <T>(): Option<T> => NONE;
```

…then adding tests that `expect(none()).toBe(none())`?

Reply:
I'm happy to make that change but couldn't this create a dual-package hazard if an application ends up with multiple instances of this library (either because it's dependency chains ends up requiring both ESM and CJS or slightly different versions)? I've been trying to avoid adding top-level state for that purpose. Is this a valid concern here?

Reply:
I've applied this suggestion but definitely think it's worth talking about it. 🙂